### PR TITLE
Bank.js

### PR DESCRIPTION
**Pull request by LaughLax**
Provides functionality for the following slash commands:
- bank give
- bank balance
- bank game list
- bank game redeem
- bank discount
- bank award

The `customId` for the command still needs to be filled in.
This PR includes updates to `config-example.json`, `snowflakes.json`, and also DB models.

Just thinking out loud, here are a couple particulars that I'd like to have assessed in the review:
- [x] English style in command descriptions (`bank.json`)
- [x] English style in messages, since I adjusted some slightly from Icarus4

Because of the change in the `Bank` DB model (renaming `mod` field to `giver`), this PR is **not** currently compatible to run alongside Icarus4 until that change is also reflected in Icarus4.

### CODE REVIEW

File: modules/bank.js [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+      };
+      let withdraw = await Module.db.bank.addCurrency(withdrawal);
+
+      interaction.reply(`You have redeemed ${gb}${withdraw.value} for a $${discount.amount} discount code in the LDS Gamers Store! <http://ldsgamers.com/shop>\n\nUse code __**${discount.code}**__ at checkout to apply the discount. This code will be good for ${discount.maxNumberOfUsages} use. (Note that means that if you redeem a code and don't use its full value, the remaining value is lost.)\n\nYou now have ${gb}${balance.balance - withdraw.value}.`);
+      let embed = u.embed()
+      .setAuthor(interaction.member.displayName, interaction.member.displayAvatarURL({dynamic: true}))
+      .addField("Amount", `${gb}${withdraw.value}\n$${withdraw.value / 100}`)
+      .addField("Balance", `${gb}${balance.balance - withdraw.value}`)
+      .setDescription(`**${Util.escapeMarkdown(interaction.member.displayName)}** just redeemed ${gb} for a store coupon code.`);
+      interaction.client.channels.cache.get(Module.config.channels.modlogs).send({embeds: embed});
```
Comment (Gaiwecoor) [inline modules/bank.js:272]:
```suggestion
      interaction.client.channels.cache.get(Module.config.channels.modlogs).send({embeds: [embed]});
```

File: modules/bank.js [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+    };
+    let receipt = await Module.db.bank.addCurrency(award);
+    let gbBalance = await Module.db.bank.getBalance(recipient.id, "gb");
+    let emBalance = await Module.db.bank.getBalance(recipient.id, "em");
+    let embed = u.embed()
+    .setAuthor(interaction.client.user.username, interaction.client.user.displayAvatarURL({dynamic: true}))
+    .addField("Reason", reason)
+    .addField("Your New Balance", `${gb}${gbBalance.balance}\n${ember}${emBalance.balance}`)
+    .setDescription(`${Util.escapeMarkdown(giver.displayName)} just ${value > 0 ?"awarded" : "docked"} you ${ember}${receipt.value}! This counts toward your House's Points.`);
+    recipient.send({embeds: embed});
```
Comment (Gaiwecoor) [inline modules/bank.js:316]:
```suggestion
    recipient.send({embeds: [embed]});
```

File: modules/bank.js [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+    interaction.reply(`${ember}${value} ${value > 0 ?"awarded to" : "docked from"} ${Util.escapeMarkdown(recipient.displayName)} for ${reason}`).then(u.clean);
```
Comment (Gaiwecoor) [inline modules/bank.js:318]: I'll have to look at `u.clean` - I forget if it will handle cleaning interaction replies.

File: modules/bank.js [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+    
+    embed = u.embed()
+    .setAuthor(interaction.client.user.username, interaction.client.user.displayAvatarURL({dynamic: true}))
+    .addField("Reason", reason)
+    .setDescription(`You just gave ${ember}${receipt.value} to ${Util.escapeMarkdown(recipient.displayName)}. This counts toward their House's Points.`);
+    giver.send({embeds: embed});
```
Comment (Gaiwecoor) [inline modules/bank.js:324]:
```suggestion
    giver.send({embeds: [embed]});
```

File: modules/bank.js [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    
+    let hoh = interaction.client.channels.cache.get(Module.config.channels.headsofhouse);
+    embed = u.embed()
+    .setAuthor(interaction.client.user.username, interaction.client.user.displayAvatarURL({dynamic: true}))      
+    .addField("Reason", reason)
+    .setDescription(`**${Util.escapeMarkdown(giver.displayName)}** ${value > 0 ?"awarded" : "docked"} ${Util.escapeMarkdown(recipient.displayName)} ${ember}${value}.`);
+    hoh.send({embeds: embed});
```
Comment (Gaiwecoor) [inline modules/bank.js:331]:
```suggestion
    hoh.send({embeds: [embed]});
```

**Review by Gaiwecoor — COMMENTED** [review]
Most comments come from how D.js 13 handles embeds. A couple items for additional discussion.

File: modules/bank.js [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+}
+
+function generateCode(n) {
+  let chars = "123456789ABCDEFGHJKLMNPQRSTUVWXYZ";
+  let newCode = "";
+  for (var i = 0; i < n; i++) {
+    newCode += u.rand(chars);
+  }
+  return newCode;
+}
```
[the remark this answers is not in the document]
Comment (LaughLax) [reply modules/bank.js:31]: I support that. Even if collisions are unlikely when codes are generated this way, I imagine `nanoid` will have ways to make them even more unlikely.

File: modules/bank.js [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+const Module = new Augur.Module()
+.addInteractionCommand({
+  name: "Bank",
+  guildId: config.ldsg,
+  commandId: "tbd",
+  process: async (interaction) => {
+    switch(interaction.options.getSubcommand(true)) {
+      case "give":
+        await bankGive(interaction);
+        break;
```
[the remark this answers is not in the document]
Comment (LaughLax) [reply modules/bank.js:344]: For a moment there I thought you meant the commands, and I was like "wait what?" The conditionals are definitely easier to read as-is, but checking based on `subCommand()` will probably not complicate it very much... I'll have to take a look and get back on this.

File: modules/bank.js [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+    // Reply so there's no "interaction failed" error message.
+    interaction.reply({content: `Watch your DMs for a list of games that can be redeemed with ${gb}!`, ephemeral: true});
+    
+    let embed = u.embed()
+    .setTitle("Games Available to Redeem")
+    .setDescription(`Redeem ${gb} for game codes with the \`!gameredeem code\` command.`);
+
+    let i = 0;
+    for (const game of games) {
+      if (((++i) % 25) == 0) {
```
[the remark this answers is not in the document]
Comment (LaughLax) [reply modules/bank.js:155]: Sounds good to me, so long as character limits are enforced per embed and not per entire message. I don't know the answer on that, but you might.

File: modules/bank.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (LaughLax) [reply modules/bank.js]: Replaced!

File: modules/bank.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (LaughLax) [reply modules/bank.js:149]: So here are the limits (relevant one in bold):
- Embed titles are limited to 256 characters
- Embed descriptions are limited to 4096 characters
- There can be up to 25 fields
- A field's name is limited to 256 characters and its value to 1024 characters
- The footer text is limited to 2048 characters
- The author name is limited to 256 characters
- **The sum of all characters from all embed structures in a message must not exceed 6000 characters**
- Ten embeds can be sent per message

So the question is, how many embeds could we send and stay within that limit?

Every field has...
- 24 characters for :gb:
- 2+ characters for amount
- 27+ characters for `/bank game redeem [CODE]`
- >58 characters for Steam link, if applicable

With 25 fields, that gives a *lower bound* estimate of 1,325 characters per embed with *zero* Steam links.

The `MessageEmbed` object *does* have a `length` attribute we could use, if we build all the embeds, set them in an array, and then dynamically decide how many to send at once. If we're going to send more than one at once, this is how I recommend it.

[the remark this answers is not in the document]
Comment (Gaiwecoor) [reply modules/bank.js:149]: Good call. Putting them in an array, then sending isn't all that different than sending as we go.

**Review by Gaiwecoor — APPROVED** [review]

**Pull request merged**